File: c/happynumbers.c
#include <stdio.h>
#include <math.h>

#define CACHE 256
enum { h_unknown = 0, h_yes, h_no };
unsigned char buf[CACHE] = {0, h_yes, 0};

//function declarations
int isHappy(int);
void fillArray(int,int);
void filter();
void swap(int *, int *);
void doubleSwap(double*,double*);
void bubbleSort(int *, int);
void normBubbleSort(double *, int);
void sortNorms();
double norm(int);

//globals
int numItems = 0, numHappies = 0;
int nums[1000], happies[1000];
double norms[1000];

// Asks the user for two positive integers
// and returns the 10 happy numbers between them
// with the greatest norms, sorted descending by the norms

// by Nate Jackson for CSC330

int main(){
	int x, y;
    printf("First Argument: ");
	scanf("%d",&x);
	printf("Second Argument: ");
	scanf("%d",&y);

	if (x == y){
		printf("NOBODYS HAPPY\n");
		return 1;
	}
	
	fillArray(x,y);
	filter();
	if (numHappies == 0){
		printf("NOBODYS HAPPY\n");
		return 1;
	}
	sortNorms();

	for (int i=0; i<10 && i<numHappies; i++) printf("%d\n",happies[i]);
	return 0;
}

//fills array with numbers between entered values
void fillArray(int x, int y){
	if (x > y){
		int i = x;
		x = y;
		y = i;
	}

	numItems = 0;
	for (int n = x; n <= y; n++){
		nums[numItems] = n;
		numItems++;
	}
}

//removes numbers from an array that are not happy
void filter() {
	numHappies = numItems;
	for (int i = 0; i < numItems; i++){
		if (!isHappy(nums[i])){
			nums[i] = 0;
			numHappies--;
		}
	}
	bubbleSort(nums, numItems);
	int numTrash = numItems-numHappies;
	for (int i = 0; i < numHappies; i++){
		happies[i] = nums[numTrash];
		numTrash++;
	}

}

//calls the norm of each happy number
//bubblesorts the happies by their norms
void sortNorms(){
	float n;
	for (int i = 0; i < numHappies; i++)
	{
		n = norm(happies[i]);
		norms[i] = n;
	}
	normBubbleSort(norms, numHappies);
}

//calculates the norm of a happy number
double norm(int n){
	int cache[1000], d = 0, cacheSize = 0;;
	double output = 0.0;
	while (n > 1){
		cache[cacheSize]=n;
		cacheSize++;
		int m = 0;
		while (n > 0){
			d = n % 10;
			m += d*d;
			n /= 10;
		}
		n = m;
	}
	
	for (int i = 0; i < cacheSize; i++){
		int j = cache[i] * cache[i];
		output += j;
	}
	output = sqrt(output);
	return output;
}

//determines if a number is happy
int isHappy(int n)
{
	int sum = 0, x, nn;
	if (n < CACHE) {
		if (buf[n]) return 2 - buf[n];
		buf[n] = h_no;
	}

	for (nn = n; nn; nn /= 10) x = nn % 10, sum += x * x;

	x = isHappy(sum);
	if (n < CACHE) buf[n] = 2 - x;
	return x;
}

//from https://www.geeksforgeeks.org/c-program-for-bubble-sort/
void swap(int *xp, int *yp){
	int temp = *xp;
	*xp = *yp;
	*yp = temp;
}
//modified swap() to take doubles
void doubleSwap(double *xp, double *yp){
	double temp = *xp;
	*xp = *yp;
	*yp = temp;
}
// A function to implement bubble sort
//from https://www.geeksforgeeks.org/c-program-for-bubble-sort/
void bubbleSort(int arr[], int n){
	int i, j;
	for (i = 0; i < n-1; i++)
		// Last i elements are already in place
		for (j = 0; j < n-i-1; j++)
			if (arr[j] > arr[j+1]) swap(&arr[j], &arr[j+1]);
}

//same as bubbleSort(), but sorts the happies array
//in tandem with the argument norm array
void normBubbleSort(double arr[], int n){
	int i, j;
	for (i = 0; i < n-1; i++)
		for (j = 0; j < n-i-1; j++)
			if (arr[j] < arr[j+1]) {
				doubleSwap(&arr[j], &arr[j+1]);
				swap(&happies[j], &happies[j+1]);
			}
}

File: go/happynumbers.go
package main

import (
	"fmt"
	"math"
	"sort"
)

// Asks the user for two positive integers
// and returns the 10 happy numbers between them
// with the greatest norms, sorted descending by the norms

// by Nate Jackson for CSC330

func main() {
	var x, y int

	fmt.Print("First Argument: ")
	fmt.Scanln(&x)
	fmt.Print("Second Argument: ")
	fmt.Scanln(&y)

	if x == y {
		fmt.Println("NOBODYS HAPPY")
		return
	}
	sortNorms(filter(createArray(x, y)))

}

//filles a slice with numbers between entered values
func createArray(x int, y int) []int {
	nums := []int{}
	if x > y {
		n := x
		x = y
		y = n
	}

	for i := x; i < y; i++ {
		nums = append(nums, i)
	}
	return nums
}

//returns a slice with numbers that are not happy removed
func filter(nums []int) []int {
	output := []int{}
	for _, i := range nums {
		if happy(i) {
			output = append(output, i)
		}
	}
	return output
}

//determines whether a number is happy
func happy(n int) bool {
	m := make(map[int]bool)
	for n > 1 {
		m[n] = true
		var x int
		for x, n = n, 0; x > 0; x /= 10 {
			d := x % 10
			n += d * d
		}
		if m[n] {
			return false
		}
	}
	return true
}

//takes in the list of happy numbers, calculates each's norm,
//sorts the list based on the norm, then prints the list of happy numbers
func sortNorms(nums []int) {
	if len(nums) == 0 {
		fmt.Println("NOBODYS HAPPY")
		return
	}
	normMap := make(map[int]float64)
	for _, i := range nums {
		n := norm(i)
		normMap[i] = n
	}
	keys := make([]int, 0, len(normMap))
	for k := range normMap {
		keys = append(keys, k)
	}
	sort.SliceStable(keys, func(i, j int) bool {
		return normMap[keys[i]] > normMap[keys[j]]
	})

	//limiting output to top 10 values
	for j := 0; j < 10 && j < len(keys); j++ {
		fmt.Println(keys[j])
	}
}

//calculates the norm of a given happy number
func norm(num int) float64 {
	cache := make(map[int]bool)

	norm := 0.0
	for num > 1 {
		cache[num] = true
		var x int
		for x, num = num, 0; x > 0; x /= 10 {
			d := x % 10
			num += d * d
		}
	}
	keys := make([]int, 0, len(cache))
	for k := range cache {
		keys = append(keys, k)
	}
	for _, x := range keys {
		y := x
		y *= y
		norm += float64(y)
	}
	norm = math.Sqrt(norm)
	return norm
}

//contains method for slice/array
func contains(sl []int, name int) bool {
	for _, value := range sl {
		if value == name {
			return true
		}
	}
	return false
}
